cdap-ui/app/cdap/components/SystemServicesDelay/index.tsx

```diff
@@ -0,0 +1,173 @@
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+import * as React from 'react';
+import { connect, Provider } from 'react-redux';
+import SystemDelayStore from 'services/SystemDelayStore';
+import SystemDelayActions from 'services/SystemDelayStore/SystemDelayActions';
+import Snackbar from '@material-ui/core/Snackbar';
+import Button from '@material-ui/core/Button';
+import ee from 'event-emitter';
+import { WINDOW_ON_FOCUS, WINDOW_ON_BLUR } from 'services/WindowManager';
+import { getExperimentValue, isExperimentEnabled } from 'services/helpers';
+import DataSource from 'services/datasource';
+import flatten from 'lodash/flatten';
+
+interface IBindingsMap {
+  [key: string]: number;
+}
+
+interface ISystemDelayProps {
+  showDelay: boolean;
+  activeDataSources: DataSource[];
+}
+
+interface ISystemDelayState {
+  cleanChecksNeeded: number;
+}
+
+const EXPERIMENT_ID = 'system-delay-notification';
+const HEALTH_CHECK_INTERVAL = 12000;
+const DEFAULT_DELAY_TIME = 5000;
+const CLEAN_CHECK_COUNT = 3;
+
+class SystemServicesDelayView extends React.Component<ISystemDelayProps> {
+  public state: ISystemDelayState = {
+    cleanChecksNeeded: 0,
+  };
+  private healthCheckInterval: NodeJS.Timeout;
+  private eventEmitter = ee(ee);
+
+  public componentDidMount() {
+    if (isExperimentEnabled(EXPERIMENT_ID)) {
+      this.checkForDelayedBindings();
+    }
+    this.startHealthCheck();
+    this.eventEmitter.on(WINDOW_ON_FOCUS, () => {
+      this.startHealthCheck();
+    });
+    this.eventEmitter.on(WINDOW_ON_BLUR, () => {
+      this.stopHealthCheck();
+    });
+  }
+
+  public componentWillUnmount() {
+    this.stopHealthCheck();
+  }
+
+  private startHealthCheck = () => {
+    if (isExperimentEnabled(EXPERIMENT_ID)) {
+      this.healthCheckInterval = setInterval(this.checkForDelayedBindings, HEALTH_CHECK_INTERVAL);
+    }
+  };
+
+  private checkForDelayedBindings = () => {
+    const delayedTimeFromExperiment = getExperimentValue(EXPERIMENT_ID);
+    const SERVICES_DELAYED_TIME = delayedTimeFromExperiment
+      ? parseInt(delayedTimeFromExperiment, 10) * 1000
+      : DEFAULT_DELAY_TIME;
+    const bindingsMap = {};
+    this.props.activeDataSources.forEach((dataSource: DataSource) => {
+      const bindings = dataSource.getBindingsForHealthCheck();
+      Object.keys(bindings).forEach((id) => {
+        bindingsMap[id] = bindings[id];
+      });
+    });
+    const currentTime = Date.now();
+    const isBindingDelayed = (id: string, bindingsMap: IBindingsMap) => {
+      const bindingStartTime = bindingsMap[id];
+      return bindingStartTime && currentTime - bindingStartTime > SERVICES_DELAYED_TIME;
+    };
+    const hasDelayedBinding = Object.keys(bindingsMap).some((id: string) =>
```

Comment:
you can combine this with the forEach at line 85. You shouldn't need to create another map, you can simply check whether the binding is delayed there and set a boolean flag whether something is delayed or not.

Reply:
Fixed